concent_api/middleman/middleman_server.py

```diff
@@ -108,26 +121,44 @@ def _close_middleman(self) -> None:
 
     async def _handle_concent_connection(self, reader, writer) -> None:
         try:
-            data = await reader.readuntil()
+            data = await handle_frame_receive_async(reader)
             remote_address = writer.get_extra_info('peername')
             print("Received %r from %r" % (data, remote_address))
-            await self._respond_to_user(data, writer)
+            serialized_message = self._prepare_response(data)
+            await self._respond_to_user(serialized_message, writer)
+        except (BrokenEscapingInFrameMiddlemanProtocolError, asyncio.LimitOverrunError) as exception:
+            await self._send_immediate_error(writer, exception)
         except Exception as exception:  # pylint: disable=broad-except
             crash_logger.error(
                 f"Exception occurred: {exception}, Traceback: {traceback.format_exc()}"
             )
             raise
 
+    @staticmethod
+    def _prepare_response(data):
+        try:
+            AbstractFrame.deserialize(data, settings.CONCENT_PUBLIC_KEY)
+        except MiddlemanProtocolError as exception:
+            error_code = map_exception_to_error_code(exception)
+            data = ErrorFrame(
+                (error_code, str(exception)),
+                0,
+            ).serialize(settings.CONCENT_PRIVATE_KEY)
```

Comment:
For consistency, `serialize` should be done in `send_over_stream_async`, like it is done in `send_over_stream`.